src/cmap/conn/stream.rs

```diff
@@ -38,27 +38,53 @@ pub(super) enum Stream {
     ),
 }
 
+fn try_connect(address: &SocketAddr, timeout: Duration) -> Result<TcpStream> {
+    // The URI options spec requires that the default connect timeout is 10 seconds, but that 0
+    // should indicate no timeout.
+    let stream = if timeout == Duration::from_secs(0) {
+        TcpStream::connect(address)?
+    } else {
+        TcpStream::connect_timeout(address, timeout)?
+    };
+
+    Ok(stream)
+}
+
+fn connect_stream(address: &StreamAddress, connect_timeout: Option<Duration>) -> Result<TcpStream> {
+    let timeout = connect_timeout.unwrap_or(DEFAULT_CONNECT_TIMEOUT);
+
+    let mut socket_addrs: Vec<_> = address.to_socket_addrs()?.collect();
+
+    if socket_addrs.is_empty() {
+        return Err(ErrorKind::NoDnsResults(address.clone()).into());
+    }
+
+    socket_addrs.sort_by_key(|addr| if addr.is_ipv4() { 0 } else { 1 });
+
+    let mut connect_error = match try_connect(&socket_addrs[0], timeout) {
+        Ok(stream) => return Ok(stream),
+        Err(err) => err,
+    };
+
+    for address in &socket_addrs[1..] {
```

Comment:
will this panic if `socket_addrs` only has one element?

Reply:
Nope; it would just yield an empty slice, meaning no iterations are done. See [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bf54f063b1349ed8139e8f65850f6596) for a playground showing this.